exercises/src/main/scala/fpinscala/datastructures/List.scala

```diff
@@ -49,20 +51,224 @@ object List { // `List` companion object. Contains functions for creating and wo
+  }
+
+  def append2[A](a1: List[A], a2: List[A]): List[A] = foldRight(a1, a2)(Cons(_ , _))
+
+  def concatenate[A](ls: List[A]*): List[A] = {
+    if (ls.isEmpty) Nil
+    else List.append2(ls.head, concatenate(ls.tail: _*))
+  }
 
-  def tail[A](l: List[A]): List[A] = sys.error("todo")
+  // def add1[A](l: List[A]): List[A] = l match {
+  //   case Nil => Nil
+  //   case Cons(x,xs) => foldRight()
+  // }
 
-  def setHead[A](l: List[A], h: A): List[A] = sys.error("todo")
+  def add1[A](l: List[Int]): List[Int] = foldRight(l, List[Int]())((x, y) => Cons(x + 1, y))
+ 
+  def listDblToStr[A](l: List[Double]): String = foldLeft(l, "")((acc, a) => acc + "*" + a.toString)
 
-  def drop[A](l: List[A], n: Int): List[A] = sys.error("todo")
+  def listDblToListStr[A](l: List[Double]): List[String] = foldRight(l, List[String]())((x, xs) => Cons(x.toString, xs))
+  
+  def map[A,B](l: List[A])(f: A => B): List[B] = foldRight(l, List[B]())((h,t) => Cons(f(h), t))
 
-  def dropWhile[A](l: List[A], f: A => Boolean): List[A] = sys.error("todo")
+  def filter[A](as: List[A])(f: A => Boolean): List[A] = foldRight(as, List[A]())((h,t) => {if (f(h)) Cons(h, t) else t})
 
-  def init[A](l: List[A]): List[A] = sys.error("todo")
+  def flatMap[A,B](l: List[A])(f: A => List[B]): List[B] = foldRight(l, List[B]())((h, t) => append2(f(h), t))
 
-  def length[A](l: List[A]): Int = sys.error("todo")
+  //def flatMap2[A,B](l: List[A])(f: A => List[B]): List[B] = concatenate(map(l)(f))
 
-  def foldLeft[A,B](l: List[A], z: B)(f: (B, A) => B): B = sys.error("todo")
+  def filter2[A](l: List[A])(f: A => Boolean): List[A] = flatMap(l)(x => if (f(x)) Cons(x, Nil) else List[A]())
+
+  def addL(a: List[Int], b: List[Int]): List[Int] = a match {
+    case Nil => Nil
+    case Cons(x, xs) => b match {
+      case Nil => Nil
+      case Cons(y, ys) => Cons(x+y, addL(xs, ys))
+    } 
+  }
+
+  def zipWith[A,B](a: List[A], b: List[A])(f: (A, A) => B): List[B] = (a, b) match {
+    case (_, Nil) => Nil
+    case (Nil, _) => Nil
+    case (Cons(x, xs), Cons(y, ys)) => Cons(f(x,y), zipWith(xs, ys)(f))
+  }
+
+  // @annotation.tailrec
+  // def hasSubsequence[A](sup: List[A], sub: List[A]): Boolean = (sup,sub) match {
+  //   case (_, Nil) => true
+  //   case (Nil, _) => false
+  //   case (Cons(x,xs), Cons(y,ys)) => ((x==y) && hasSubsequence(xs,ys)) || hasSubsequence(xs,sub)
+  // }
+
+  @annotation.tailrec
+  def startsWith[A](a: List[A], b: List[A]): Boolean = (a,b) match {
+    case (_,Nil) => true
+    case (Cons(x,xs), Cons(y,ys)) if (x==y) => startsWith(xs,ys)
+    case(_,_) => false 
+  }
+
+   @annotation.tailrec
+   def hasSubsequence[A](sup: List[A], sub: List[A]): Boolean = (sup,sub) match {
+     case (_,Nil) => true
+     case (Nil,_) => false
+     case (_,_) if startsWith(sup,sub) => true
+     case (Cons(x,xs), _) => hasSubsequence(xs,sub) 
+   }
 
-  def map[A,B](l: List[A])(f: A => B): List[B] = sys.error("todo")
 }
+
+/*
```

Comment:
Generally we should remove medium-to-large areas of commented-out code (if it's not part of documentation before merging a PR.